training/train_instance.py

```diff
@@ -0,0 +1,125 @@
+
+        # Ensure autogenerated_config_dir exists
+        pathlib.Path(autogenerated_config_dir).mkdir(
+            parents=True, exist_ok=True)
+
+    def __reset(self):
+        self.inactive = True
+        self.process = None
+        self.id = None
+
+    def train(self, meta_params):
+        """ Launch training instance
+        """
+        self.inactive = False
+        self.meta_params = meta_params
+        self.id = self.__dict_to_string(meta_params)
+        self.__create_config_file()
+        command = self.__get_command()
+        self.output_file = os.path.join(output_files_dir, str(self.id) + ".out")
+        log = open(self.output_file, 'a')
+        self.process = subprocess.Popen(command, stdout=log, stderr=log, shell=True)
+        print("Training started with PID: " + str(self.process.pid))
+
+    def is_done(self):
+        """ True if training has finished / converged / not promising
+        """
+        status = self.process.poll()
+        if status is not None and status == 0:
+            return True
+        # TODO: early stopping heuristics
+        return False
+
+    def kill(self):
+        """ Send SIGINT signal to process
+        """
+        print("Killing", self.process.pid)
+        try:
+            self.process.send_signal(signal.SIGINT)
+        except Exception as e:
+            print(e)
+        self.__reset()
+
+    def get_val(self):
+        """ Return training result (for the optimizer)
+        """
+        with open(self.output_file, 'r') as f:
+            lines = f.read().splitlines()
+            for i in range(50):  # Check the last 50 lines for a mean reward
+                try:
+                    last_line = lines[-i]
+                    if "Mean Reward:" in last_line:
+                        # TODO some fancy heuristic instead of last value
+                        val = float(last_line.split("Mean Reward: ")[1].split(". Std of Reward")[0])
+                        return val
+                except Exception as e:
+                    print(e)
+                    pass
+        return 0
+
+    def __create_config_file(self):
+        self.config_dir = os.path.join(
+            autogenerated_config_dir, 'config_' + str(self.id) + '.yaml')
+        # Join dictionaries
+        params = copy.deepcopy(self.template_params)
+        for key in self.meta_params.keys():
+            v = self.meta_params[key]
+            if not isinstance(v, (int, float, str)):  # Hack to convert np vars to native python
+                v = v.item()
+            params["default"][key] = v
+        with open(self.config_dir, 'w') as outfile:
+            yaml.dump(params, outfile, default_flow_style=False)
+
+    def __get_command(self):
+        # "mlagents-learn train_configs/config.yaml --env=Builds/multiple_instances.x86_64 --run-id=test_mi --time-scale=100 --no-graphics --base-port=5100 --train >> output.nohup"
+        command = "cd " + parent_dir + "; workon py3;"
```

Comment:
workon is a bit specific to your platform and virtual environment, maybe we should just call it before running the script?

Reply:
Yes, you are right. I was afraid it wouldnt work on my computer if I didnt put it, will remove it and check if it works.